prov/efa/src/rxr/rxr_pkt_type_misc.c

```diff
@@ -67,41 +67,69 @@ ssize_t rxr_pkt_init_handshake(struct rxr_ep *ep,
 	return 0;
 }
 
-void rxr_pkt_post_handshake(struct rxr_ep *ep,
-			    struct rxr_peer *peer,
-			    fi_addr_t addr)
+/*
+ * This function is used for posting handshake packet to a peer.
+ * The return value is returned to the caller to decide either to
+ * retry later or hard fail.
+ *
+ * ep: The endpoint on which the handshake packet is sent out.
+ * addr: The address of the peer to which the handshake packet is posted.
+ */
+ssize_t rxr_pkt_post_handshake(struct rxr_ep *ep, fi_addr_t addr)
 {
 	struct rxr_pkt_entry *pkt_entry;
 	ssize_t ret;
 
-	assert(!(peer->flags & RXR_PEER_HANDSHAKE_SENT));
-
 	pkt_entry = rxr_pkt_entry_alloc(ep, ep->tx_pkt_efa_pool);
 	if (OFI_UNLIKELY(!pkt_entry))
-		return;
+		return -FI_EAGAIN;
 
 	rxr_pkt_init_handshake(ep, pkt_entry, addr);
 
-	/*
-	 * TODO: Once we start using a core's selective completion capability,
-	 * post the HANDSHAKE packets without FI_COMPLETION.
-	 */
 	ret = rxr_pkt_entry_send(ep, pkt_entry, addr);
-
-	/*
-	 * Skip sending this handshake on error and try again when processing the
-	 * next REQ from this peer containing the source information
-	 */
 	if (OFI_UNLIKELY(ret)) {
 		rxr_pkt_entry_release_tx(ep, pkt_entry);
-		if (ret == -FI_EAGAIN)
-			return;
-		FI_WARN(&rxr_prov, FI_LOG_CQ,
-			"Failed to send a HANDSHAKE packet: ret %zd\n", ret);
-		return;
 	}
+	return ret;
+}
 
-	peer->flags |= RXR_PEER_HANDSHAKE_SENT;
+/*
+ * This function is used for posting handshake packet to a peer. If
+ * FI_EAGAIN happens, peer will be added into queue_peer_list for
+ * retry later. For other errors, we will hard fail.
+ *
+ * ep: The endpoint on which the handshake packet is sent out.
+ * peer: The peer to which the handshake packet is posted.
+ * addr: The address of the peer.
+ */
+void rxr_pkt_post_handshake_or_queue(struct rxr_ep *ep, struct rxr_peer *peer,
+				     fi_addr_t addr)
+{
+	ssize_t ret;
+
+	assert(!(peer->flags & RXR_PEER_HANDSHAKE_SENT_OR_QUEUED));
+
+	ret = rxr_pkt_post_handshake(ep, addr);
+	if (OFI_UNLIKELY(ret)) {
+		if (ret == -FI_EAGAIN) {
+			// add peer to peer_queued_list for retry later
+			dlist_insert_tail(&peer->queued_entry,
```

Comment:
Is `queued_entry` used to link to lists other than `ep->queued_peer_list`?

Reply:
No, it's used to list peer to `ep->queued_peer_list`.